server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -93,9 +85,11 @@
     public TransportShardBulkAction(Settings settings, TransportService transportService, ClusterService clusterService,
                                     IndicesService indicesService, ThreadPool threadPool, ShardStateAction shardStateAction,
                                     MappingUpdatedAction mappingUpdatedAction, UpdateHelper updateHelper, ActionFilters actionFilters,
-                                    IndexingPressure indexingPressure, SystemIndices systemIndices) {
+                                    IndexingPressure indexingPressure, SystemIndices systemIndices,
+                                    ExecutorSelectorService executorSelectorService) {
         super(settings, ACTION_NAME, transportService, clusterService, indicesService, threadPool, shardStateAction, actionFilters,
-            BulkShardRequest::new, BulkShardRequest::new, EXECUTOR_NAME_FUNCTION, false, indexingPressure, systemIndices);
+            BulkShardRequest::new, BulkShardRequest::new, ExecutorSelectorService::getWriteExecutorForShard, false,
+            indexingPressure, systemIndices, executorSelectorService);
```

Comment:
A lot of the classes that I've added an executorSelectorService to already have access to the SystemIndices object.

Since the executorSelectorService is only constructed with the system indices object, would it make more sense for a `SystemIndices` object to have a `getExecutorSelector` method that returns the `ExecutorSelectorService` object? It would certainly clean things up in this PR.

Reply:
I think it would be fine for the `SystemIndices` class to enable access to the service. One note is that I would rename the class `ExecutorSelector` and drop the service portion of the name.